src/streams.cpp

```diff
@@ -5,6 +5,54 @@
 #include <span.h>
 #include <streams.h>
 
+#include <array>
+
+std::size_t XorFile::detail_fread(Span<std::byte> dst)
+{
+    if (!m_file) throw std::ios_base::failure("XorFile::read: file handle is nullptr");
+    const auto init_pos{std::ftell(m_file)};
+    if (init_pos < 0) return 0;
+    std::size_t ret{std::fread(dst.data(), 1, dst.size(), m_file)};
+    util::Xor(dst.subspan(0, ret), m_xor, init_pos);
+    return ret;
+}
+
+void XorFile::read(Span<std::byte> dst)
+{
+    if (detail_fread(dst) != dst.size()) {
+        throw std::ios_base::failure{feof() ? "XorFile::read: end of file" : "XorFile::read: failed"};
+    }
+}
+
+void XorFile::ignore(size_t num_bytes)
+{
+    if (!m_file) throw std::ios_base::failure("XorFile::ignore: file handle is nullptr");
+    std::array<std::byte, 4096> buf;
+    while (num_bytes > 0) {
+        auto buf_now{Span{buf}.first(std::min<size_t>(num_bytes, buf.size()))};
+        if (std::fread(buf_now.data(), 1, buf_now.size(), m_file) != buf_now.size()) {
+            throw std::ios_base::failure{feof() ? "XorFile::ignore: end of file" : "XorFile::ignore: failed"};
+        }
+        num_bytes -= buf_now.size();
+    }
+}
+
+void XorFile::write(Span<const std::byte> src)
+{
+    if (!m_file) throw std::ios_base::failure("XorFile::write: file handle is nullptr");
+    std::array<std::byte, 4096> buf;
```

Comment:
fa9325a14584be1c5aa6b4cdefbdaa69b27e402d

Might be interesting to see how statically allocating a buffer (say, as a member of `XorFile`) and reusing it might affect performance; may avoid some reallocations. Not sure how frequently this is actually called though. I expect this probably won't make a difference though, because I guess `XorFile::write()` will be called on the order of once per block.

Reply:
I haven't run benchmarks against statically allocating a buffer, but it may be useful since `AutoFile::write` is called many times per block. Basically every time the pattern `CAutoFile << x` is used which happens for every element of the block header, each transaction input, output, etc. I also ran a small unit test through `perf` where there were many writes and found that `std::ftell` dominated with `std::fwrite` close behind.

Reply:
I think it must have been a fluke because when I synced up to a little past block ~200K, `fwrite` was in the lead and `ftell` and xor were about tied at ~3% each of the total time spent. I wrote a test here https://github.com/Crypt-iQ/bitcoin/commit/bf924d6d2d4020e799796c0751ac64c892ae8e6b that requires the benchmarks to be compiled but is not a benchmark.  I made a patch in the following commit (https://github.com/Crypt-iQ/bitcoin/commit/be07dbc6f26638f7ec187e42144ccc5dab90c0a7) that reduced the number of `ftell` calls to 1 for `AutoFile` and it gave a noticeable performance improvement.  I think it's ok to track the file position this way, but I might be wrong